lang/src/org/partiql/lang/eval/visitors/GroupByPathExpressionVisitorTransform.kt

```diff
@@ -0,0 +1,169 @@
+                expr.steps.size != 1                               -> false
+                else                                               -> true
+            }
+
+        }
+
+        /**
+         * Collects all of the aliases defined by the specified [FromSource] and its children.
+         * This is why this rewrite must occur after [FromSourceAliasRewriter].
+         */
+        fun collectAliases(fromSource: PartiqlAst.FromSource): List<String> =
+            when (fromSource) {
+                is PartiqlAst.FromSource.Scan    ->
+                    listOf(
+                        fromSource.asAlias?.text
+                        ?: errNoContext("FromSourceItem.variables.asName must be specified for this rewrite to work", internal = true))
+
+                is PartiqlAst.FromSource.Join    ->
+                    collectAliases(fromSource.left) + collectAliases(fromSource.right)
+
+                is PartiqlAst.FromSource.Unpivot ->
+                    listOfNotNull(fromSource.asAlias?.text, fromSource.atAlias?.text)
+            }
+    }
+
+    override fun transformExprSelect(node: PartiqlAst.Expr.Select): PartiqlAst.Expr {
+        val fromSourceAliases = collectAliases(node.from)
+
+        // These are substitutions for path expressions that do not contain references to shadowed variables
+        val unshadowedSubstitutions = getSubstitutionsExceptFor(fromSourceAliases)
+
+        // A rewriter for the above
+        val unshadowedRewriter = GroupByPathExpressionVisitorTransform(unshadowedSubstitutions)
+
+        // These are the substitutions originating from the GROUP BY clause of the current [Select] node.
+        val currentSubstitutions = getSubstitutionsForSelect(node)
+
+        // A rewriter for both of the sets of the substitutions defined above.
+        val currentAndUnshadowedRewriter = GroupByPathExpressionVisitorTransform(
+            unshadowedSubstitutions + currentSubstitutions)
+
+        // Now actually rewrite the query using the appropriate rewriter for each of various clauses of the
+        // SELECT statement.
+
+        val projection = currentAndUnshadowedRewriter.transformProjection(node.project)
+
+        // The scope of the expressions in the FROM clause is the same as that of the parent scope.
+        val from = this.transformFromSource(node.from)
+
+        val fromLet = node.fromLet?.let { unshadowedRewriter.transformLet(it) }
+
+        val where = node.where?.let { unshadowedRewriter.transformExprSelect_where(node) }
+
+        val groupBy = node.group?.let { unshadowedRewriter.transformGroupBy(it) }
+
+        val having = node.having?.let { currentAndUnshadowedRewriter.transformExprSelect_having(node) }
+
+        val limit = node.limit?.let { unshadowedRewriter.transformExprSelect_limit(node) }
+
+        val metas = unshadowedRewriter.transformExprSelect_metas(node)
```

Comment:
For PIG generated functions, could have called the outer function (i.e. `transformExprSelect_fromLet`) rather than the inner function (i.e. `transformLet`). Some SFW clauses like `WHERE` only have the outer clause.

Should I make these calls more consistent with each other?

Reply:
It probably doesn't matter a huge deal, but I suppose it is technically more correct to call the `transformExprSelect_*` functions instead.